addons/mail/static/tests/chatter_tests.js

```diff
@@ -633,6 +633,268 @@ QUnit.test('chatter: post a message and switch in edit mode', function (assert)
+
+            // send message
+            form.$('.oe_chatter .o_composer_button_send').click();
+
+            var $modal = $('.modal-dialog');
+            assert.strictEqual($modal.length, 1, "should have a modal opened");
+            assert.strictEqual($modal.find('.modal-body').text(),
+                "The record has been modified, your changes will be discarded. Do you want to proceed?",
+                "should warn the user that any unsaved changes will be lost");
+
+            form.destroy();
+            done();
+        });
+});
+
+QUnit.test('chatter: discard changes on opening full-composer', function (assert) {
+    // When we open the full-composer, any following operations by the user
+    // will reload the record (even closing the full-composer). Therefore,
+    // we should warn the user when we open the full-composer if the record
+    // is dirty (= has some unsaved changes).
+    assert.expect(2);
+
+    var messages = [];
+    var bus = new Bus();
+    var form = createView({
+        View: FormView,
+        model: 'partner',
+        data: this.data,
+        arch: '<form string="Partners">' +
+                '<sheet>' +
+                    '<field name="foo"/>' +
+                '</sheet>' +
+                '<div class="oe_chatter">' +
+                    '<field name="message_ids" widget="mail_thread"' +
+                        ' options="{\'display_log_button\': True, \'post_refresh\': \'always\'}"/>' +
+                '</div>' +
+            '</form>',
+        res_id: 2,
+        session: {},
+        mockRPC: function (route, args) {
+            if (route === "/web/dataset/call_kw/partner/message_get_suggested_recipients") {
+                return $.when({2: []});
+            }
+            return this._super(route, args);
+        },
+        intercepts: {
+            get_messages: function (ev) {
+                ev.stopPropagation();
+                var requested_msgs = _.filter(messages, function (msg) {
+                    return _.contains(ev.data.options.ids, msg.id);
+                });
+                ev.data.callback($.when(requested_msgs));
+            },
+            get_bus: function (ev) {
+                ev.stopPropagation();
+                ev.data.callback(bus);
+            },
+        },
+        viewOptions: {
+            mode: 'edit',
+        },
+    });
+
+    // Make record dirty
+    form.$('.o_form_sheet input').val('trululu').trigger('input');
+
+    // Open full-composer
+    form.$('.o_chatter_button_new_message').click();
+    form.$('.o_composer_button_full_composer').click();
+
+    var $modal = $('.modal-dialog');
+    assert.strictEqual($modal.length, 1, "should have a modal opened");
+    assert.strictEqual($modal.find('.modal-body').text(),
+        "The record has been modified, your changes will be discarded. Do you want to proceed?",
+        "should warn the user that any unsaved changes will be lost");
+
+    form.destroy();
+});
+
```

Comment:
i'd add a test where the confirm dialog isn't displayed

Reply:
Ok, see test 'chatter: discard changes on message post without post_refresh'